### DISCUSSION TITLE
Danfoss Ally minimum and maximum temperatures

### DISCUSSION THREAD

**Discussion by JackJensen** [Koenkk/zigbee2mqtt#11125] [Devices]
Hi,

In my Home Assistant I have just switched from using ZHA to Zigbee2mqtt and I see something strange on my Danfoss Ally thermostats.
Previously it was only possibly in Home Assistant to regulate the thermostats in the supported temperature range of 6-28°C, but after changing to zigbee2mqtt I can set the temperature between 5-32°C. 
Setting the temperature outside the supported area is ignored, but it is a little annoying that the range is larger than what the device supports.
Anyone who knows where this extended range originates from?

**Comment by JackJensen** [comment]
In case someone else experience the same issue, here is what I found.
The range is defined in [zigbee-herdsman-converters](https://github.com/Koenkk/zigbee-herdsman-converters) /devices/danfoss.js.

According to the specs. from Danfoss (that I have been able to find) the range is 5-35°C, so confusion is pretty much complete.
I solved my issue by setting the two hvacThermostat properties minHeatSetpointLimit=500 and maxHeatSetpointLimit=3200 in all my thermostats using the zigbee2mqtt Dev Console.

Now I am wondering if it would be possible to get these values from the thermostat instead of hardcoding them in the converter?

**Reply by sjorge** [reply]
If that is the case, we should probably set the min and maxHeadSetpointLimit in the configure() function for the device. And use the same one in the expose.

**Reply by JackJensen** [reply]
That sounds like a small thing to do?
If I have to do this I have to learn javascript first (I am a C, C++ and C# developer), so that will take some time, but if someone can do this, I will be happy to help with the testing.

**Reply by sjorge** [reply]
I made a PR here, I did not test it as I no longer have this device. https://github.com/Koenkk/zigbee-herdsman-converters/pull/3808

You should be able to edit your danfoss.js and restart zigbee2mqtt, there was a trick for docker users too but I don't use docker so I'm not sure how to do it there exactly.

If all goes well, reconfiguring the device should set minHeatSetpointLimit and maxHeatSetpointLimit

**Reply by JackJensen** [reply]
@sjorge Just to make sure I expect the right outcome of this.
1) When the thermostat is paired it will get it's maximum range set to 5-32°C.
2) I change the maximum range through the Dev console in the frontend to e.g. 7-30°C.
3) Press the reconfigure button on the Device page. The thermostat returns to the range being 5-32°C.

no. 1 works as I would expect and the range is forced to 5-32°C.
no. 3 has no effect as I get the following error:

```
In "await reporting.bind(endpoint, coordinatorEndpoint, ['genPowerCfg', 'hvacThermostat']);"

Zigbee2MQTT:error 2022-02-03 19:34:26: Failed to configure 'Termostat.Kontor', attempt 1 (Error: Bind 0xccccccfffe1daa87/1 genPowerCfg from '0x00212effff06de19/1' failed (waiting for response TIMEOUT)
    at DeconzAdapter.bind (C:\workspaces\zigbee2mqtt\node_modules\zigbee-herdsman\src\adapter\deconz\adapter\deconzAdapter.ts:753:19)
    at runNextTicks (node:internal/process/task_queues:61:5)
    at processTimers (node:internal/timers:497:9)
    at Endpoint.bind (C:\workspaces\zigbee2mqtt\node_modules\zigbee-herdsman\src\controller\model\endpoint.ts:517:13)
    at Object.bind (C:\workspaces\zigbee2mqtt\node_modules\zigbee-herdsman-converters\lib\reporting.js:33:9)
    at Object.configure (C:\workspaces\zigbee2mqtt\node_modules\zigbee-herdsman-converters\devices\danfoss.js:85:13)
    at Configure.configure (C:\workspaces\zigbee2mqtt\lib\extension\configure.ts:115:13)
    at Configure.onMQTTMessage (C:\workspaces\zigbee2mqtt\lib\extension\configure.ts:55:21))
```

**Reply by sjorge** [reply]
Yep that sounds correct.
Looks like the reconfigure hit a timeout, the TRV might have been sleeping so you might need to physical wake it up before doing the reconfigure

**Comment by robertalexa** [comment]
@sjorge I have seen you tagging me in the PR.

After reading the above, I don't see any issue here? I think it is a user confusion?

So the exposed attribute is currently limited to 5 to 32 C in 0.5 C increments
![image](https://user-images.githubusercontent.com/9019306/152532515-e5881eba-8e17-4182-920f-e74931fc57d5.png)


Based on documentation the accepted range is 5 to 35 C
![image](https://user-images.githubusercontent.com/9019306/152532614-1cae686e-53e8-4758-ad2b-8ccf89db1bba.png)
![image](https://user-images.githubusercontent.com/9019306/152532632-4211b12c-bec7-4e38-b58c-e963694f7669.png)

Given the default range on the default from factory is 5 to 35 (5 to 32 on my Hive TRVs, i think Popp too, but can't remember), and the expose being 5 to 32, i don't see any problems with the current behaviour? There is also no need to set the "default" via device configure as that is already the default? (e.g. i don't think your PR will achieve anything, as that is default value anyway, unless i completely miss the point)

Z2M used to be 6 to 28 a while ago, which i have corrected some time ago. Because other zigbee "gateway" software always took inspiration from z2m, that would explain why they still say 6 to 28. But 5 to 32 (35) is actually a valid range for the Danfoss.

If i understand OP correctly, he wants to restrict that even further? The following statement confuses me a lot :)


> Setting the temperature outside the supported area is ignored, but it is a little annoying that the range is larger than what the device supports.

**Reply by robertalexa** [reply]
Later thoughts, I wonder if OP set the device in pairing mode but didn't factory reset them when transitioning from ZHA to Z2M, and the devices might have kept "old" configurations from ZHA (6 to 28) - since they are firmware level limits

**Reply by sjorge** [reply]
That's possible, or it could be the Danfoss out of the box has a more limited range. But I no longer have access to the device to check. In the later case the change will set the max range when reconfigured.

**Reply by robertalexa** [reply]
I mean it is possible.

My judgement is, considering Danfoss is the only one out of the 3 (Danfoss, Hive, Popp) that are super transparent about their devices, i kind trust their devices to leave the factory with the right settings, that combined withe the fact you can download the most recent firmware, makes me think that it is highly unlikely to be wrong.

I mean your PR can't harm (except when people set custom ones over 10 radiators, and then the device gets reconfigured for some reason, so the user needs to do all of them again).

I will try to test this on my Hives and comment on the PR

**Comment by JackJensen** [comment]
It was never my intention to cause confusion.
1) I did not do a factory reset at first, so you are most likely right that my settings are from ZHA (6-28°C), but the Z2M and Danfoss default settings does not match either, so I will be able to set 5-35 on the thermostat, but HA will only show 5-32°C.
2) "Setting the temperature outside the supported area is ignored, but it is a little annoying that the range is larger than what the device  supports." 
While I had the ZHA (6-28) settings in the thermostat and Z2M tried to set the temperature to 5°C, the thermostat would just ignore it.
HA would show the range reported by Z2M (5-32), but the thermostat could only be set between 6 and 28. So if I tried to change the temperature from 21°C to 5°C, HA would show 5°C, but the thermostat would stay at 21°C.
3) Factory resetting will set the range to 5-35°C
4) My initial thought on how this should be solved was to have expose use the values from the thermostat, instead of forcing hardcoded values on the thermostat during configuration. I don't know if that is possible, but that would also solve the problem with users having custom ranges and reconfiguring the thermostats.

**Reply by robertalexa** [reply]
Hi Jack, thanks a lot for clarifying, i think i now fully understand what you meant.

While i perfectly agree with the 5-32 vs 5-35 range, the reason for 5-32 range was that this way the Hive TRV and Popp also operate correctly. They are physically the same device, but bespoke firmware for those 2 sellers. The compromise was accepted historically because in practice people will not really set their room temperature to 32 (or 35 for that matter). On the Hive TRV you can't even increase the hardware limit to 35, the max allowed value is 32. So whichever way we play it, one set of devices losses.

I mean for correctness, we can increase the limit to 35, and have Hive and Popp "suffer" if a user tries to go that high. But realistically it probably makes no difference to either set of users.

While i understand your logic at point 4, that is not possible due to the nature of z2m, we have had a similar debate a while ago about the number of endpoints available for Danfoss Icon floor heating device, and we had to make a compromise.

With regards to your point 2, that is a result of using "Optimistic" which set the value without receiving confirmation from the device. Without optimistic, z2m will not change its value until the device confirms (which in your case would have failed with old settings), and in turns HA would not be updated by z2m either.

I hope the above answers your questions, let me know if i can help any further

**Reply by JackJensen** [reply]
@[robertalexa](https://github.com/robertalexa) Thanks for your answer.

5-32 vs 5-35 range is not important. I was looking for consistency.

> With regards to your point 2, that is a result of using "Optimistic" which set the value without receiving confirmation from the device. Without optimistic, z2m will not change its value until the device confirms (which in your case would have failed with old settings), and in turns HA would not be updated by z2m either.

That makes sense. Thank you.

Given that it is not possible to use the settings from the device in expose, I don't think it makes sense to change anything (except that it would ensure that the range in expose and device consistent after a configuration)

**Reply by robertalexa** [reply]
Unfortunately not even after reconfiguration you can't change the exposed attribute range.

However there is a workaround, and i think this is what you want to achieve. Here is how i approached it

- Z2M ranges from 5 to 32
- manually set the trv min setpoint to 9C (900)
- the expose range is still 5 to 32 but i don't care because i don't change it from z2m, i use HA
- but HA also now used 5 to 32 because of the expose
- customize each TRV, this is the solution :)

https://www.home-assistant.io/docs/configuration/customizing-devices/

In configuration.yaml
![image](https://user-images.githubusercontent.com/9019306/152549713-6e7feccd-bff7-44e5-a09f-3a2d18b2756a.png)

In customize.yaml
![image](https://user-images.githubusercontent.com/9019306/152549797-34880794-18f0-4ffb-98cf-fd84f9bac199.png)
Use max_temp if you also change the max, i don't. Google Home (assistant) supports a range of 9 to 32, so for how z2m works, 32 matches, and makes everything consistent.

Go to Setting -> Controls, reload customizations
![image](https://user-images.githubusercontent.com/9019306/152549896-850bb412-de61-4a2f-ba7e-0305142433a1.png)

The temperature sliders will update within minutes as soon as there is an update from z2m.

So in short, you will not be able to set outside of the range via HA, you will via z2m, but you probably won't ever do that.

PS: i also have a Node-Red fallback, to auto change the temp if lower than 9. This is because I have never bothered looking into changing the min setpoint for my Hive SLR1. who knows, that might work and I might not even need the fallback. But i am keeping it in case my above setup ever fails, and you end up physically changing the device, maybe after a factory reset or sth.

![image](https://user-images.githubusercontent.com/9019306/152550921-79b4041a-0fd6-425d-ad8f-6cf4968dd2de.png)


I hope this above helps you achieve the consistency you are after :)

LE: i just changed my min heat setpoint on Hive SLR1 (main boiler thermostat), that works, so now I can technically get rid of the Node-Red fallback. I will still probably keep it as it won't hurt anyone

**Reply by robertalexa** [reply]
Further note: once you set the min to 9C, even if you try to set it to 5 via z2m, the temp will revert to 9 automatically, so z2m is also "safe".
